File: check_alignment.dart
import 'dart:io';
import 'dart:typed_data';

void main() async {
  final apkPath = 'build/app/outputs/flutter-apk/app-release.apk';
  final file = File(apkPath);

  if (!await file.exists()) {
    print('Error: $apkPath not found.');
    return;
  }

  final bytes = await file.readAsBytes();
  final buffer = bytes.buffer.asUint8List();

  print('${"Data Offset".padRight(15)} ${"Path"}');
  print("-" * 60);

  // Simple ZIP parser to find local headers
  // Local header signature: 0x04034b50
  for (int i = 0; i < buffer.length - 30; i++) {
    if (buffer[i] == 0x50 && buffer[i+1] == 0x4b && buffer[i+2] == 0x03 && buffer[i+3] == 0x04) {
      final data = buffer.sublist(i + 26, i + 30);
      final nameLen = ByteData.sublistView(data, 0, 2).getUint16(0, Endian.little);
      final extraLen = ByteData.sublistView(data, 2, 4).getUint16(0, Endian.little);
      
      if (i + 30 + nameLen <= buffer.length) {
        final name = String.fromCharCodes(buffer.sublist(i + 30, i + 30 + nameLen));
        
        if (name.endsWith('.so')) {
          final dataOffset = i + 30 + nameLen + extraLen;
          final alignment = dataOffset % 16384;
          final status = alignment == 0 ? "✅ ALIGNED" : "❌ NOT ALIGNED";
          print('${dataOffset.toString().padRight(15)} $name (mod 16384 = $alignment) $status');
        }
      }
    }
  }
}

File: check_alignment.py
import zipfile
import os

apk_path = r'build\app\outputs\flutter-apk\app-release.apk'

def check_alignment(apk_path):
    if not os.path.exists(apk_path):
        print(f"Error: {apk_path} not found.")
        return

    with open(apk_path, 'rb') as f:
        z = zipfile.ZipFile(f)
        print(f"{'Alignment':<10} {'Path'}")
        print("-" * 50)
        for info in z.infolist():
            if info.filename.endswith('.so'):
                # Header offset in the ZIP
                offset = info.header_offset
                # The data starts after the local header
                # We need to find the actual start of the compressed/stored data
                # A quick way is to read the local header size
                f.seek(offset)
                local_header = f.read(30)
                n = int.from_bytes(local_header[26:28], 'little') # name length
                m = int.from_bytes(local_header[28:30], 'little') # extra length
                data_offset = offset + 30 + n + m
                
                alignment = data_offset % 16384
                print(f"{data_offset:<10} {info.filename} (mod 16384 = {alignment})")

if __name__ == "__main__":
    check_alignment(apk_path)
